src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,1209 @@
+                    }
+                    410 | 451 => {
+                        // The crate was deleted from the registry.
+                        // Errors will eventually be yielded by load().
+                        todo!("we should delete the local index file here if it exists");
+                    }
+                    code => {
+                        anyhow::bail!(
+                            "prefetch: server returned unexpected HTTP status code {} for {}{}",
+                            code,
+                            self.source_id.url(),
+                            fetched.path.display()
+                        );
+                    }
+                }
+            }
+
+            if !self.downloads.eager.is_empty() {
+                continue;
+            }
+
+            if self.downloads.pending.is_empty() {
+                // We're all done!
+                break;
+            }
+
+            // We have no more replies to provide the caller with,
+            // so we need to wait until cURL has something new for us.
+            let timeout = self
+                .prefetch
+                .get_timeout()?
+                .unwrap_or_else(|| Duration::new(5, 0));
+            self.prefetch
+                .wait(&mut [], timeout)
+                .chain_err(|| "failed to wait on curl `Multi`")?;
+        }
+
+        debug!("prefetched all transitive dependencies");
+        self.is_prefetching = false;
+        Ok(None)
+    }
+
+    fn index_path(&self) -> &Filesystem {
+        // NOTE: I'm pretty sure this method is unnecessary.
+        // The only place it is used is to set `.path` in `RegistryIndex`,
+        // which only uses it to call `assert_index_locked below`...
+        &self.index_path
+    }
+
+    fn assert_index_locked<'a>(&self, path: &'a Filesystem) -> &'a Path {
+        self.config.assert_package_cache_locked(path)
+    }
+
+    fn current_version(&self) -> Option<InternedString> {
+        // TODO: Can we use the time of the last call to update_index here?
+        None
+    }
+
+    fn update_index_file(&mut self, root: &Path, path: &Path) -> CargoResult<bool> {
+        let pkg = root.join(path);
+        if pkg.exists() {
+            paths::remove_file(&pkg)?;
+        }
+        // Also reset self.fresh so we don't hit an assertion failure if we re-download.
+        self.fresh.remove(path);
+        Ok(true)
+    }
+
+    fn load(
+        &mut self,
+        root: &Path,
+        path: &Path,
+        data: &mut dyn FnMut(&[u8]) -> CargoResult<()>,
+    ) -> CargoResult<()> {
+        // NOTE: This is pretty much a synchronous version of the prefetch() + next_prefetched()
+        // dance. Much of the code is sort-of duplicated, which isn't great, but it's moderalyte
+        // straightforward and works. When the real resolver supports a load returning "not yet",
+        // load and prefetch can be merged.
```

Comment:
FWIW this is definitely something that will need to get deduplicated in the long term. There's a lot of nontrivial logic in this function duplicated with the above prefetching.

Reply:
Yes, I agree. I believe your proposed resolver changes will get rid of this duplication, since "prefetching" will go away and load will become asynchronous. My proposal would be to resolve this for now.